### PR TITLE
allow plugins to register daemon commands using decorator

### PR DESCRIPTION
**Pull request by markblundeberg**
While we can't add new @commands in plugins (since it's too late), we can have the daemon RPC accept arbitrary subcommands and dispatch them to plugins, when desired.

Note in order to get this working on the non-GUI daemon, you have to add a `cmdline.py` to your plugin with a GUI-free implementation of the `class Plugin:`, then put `available_for = ['qt', 'cmdline']` in your plugin's `__init__.py`. In principle also you should do this for `stdio` and `text` which are alternative non-graphical GUIs; each of these can just be short stub files.

I've tested this on a sample plugin and it works well.

~~The usage of metaclass is inspired by this: https://stackoverflow.com/a/3054505/5667904~~
The usage of scan-class-on-init is inspired by  #1669

### CODE REVIEW

**Comment by cculianu** [comment]
Good guy Mark. Works on CashFusion and decides to add proper CLI support. Nice.  I will review this very soon. 👍

File: lib/plugins.py [hunk 1]
```diff
@@ -545,6 +572,15 @@ def __init__(self, parent, config, name):
                 hooks[aname].append((self, func))
                 self._hooks_i_registered.append((aname,func))
 
+        # collect names of all class attributes with ._is_daemon_command
+        self._daemon_commands = tuple(attrname for attrname in dir(type(self))
```
Comment (cculianu) [inline lib/plugins.py]: Hmm. Why save `self._daemon_commands` to an instance attribute? It seems to not be accessed outside this function.. couldn't a temp var just work fine here?

If you plan on adding commits though, that make use of this instance dict: may I suggest at least clearing it on `.close()` (see below)..?

File: lib/plugins.py [hunk 2]
```diff
@@ -564,6 +600,12 @@ def close(self):
             if not l:
                 hooks.pop(name, None)
         self._hooks_i_registered.clear()  # just to kill strong refs to self ASAP, for GC
+        # remove registered daemon commands
+        for cmdname in self._daemon_commands:
+            try:
+                del self.parent.daemon_commands[k]
+            except KeyError:
+                pass # again, shouldn't happen
```
Comment (cculianu) [inline lib/plugins.py]: Why not also do `self._daemon_command.clear()` here when done?  (Again, for same reason it was done above -- GC purposes, etc).

File: lib/plugins.py [hunk 3]
```diff
@@ -564,6 +600,12 @@ def close(self):
             if not l:
                 hooks.pop(name, None)
         self._hooks_i_registered.clear()  # just to kill strong refs to self ASAP, for GC
+        # remove registered daemon commands
+        for cmdname in self._daemon_commands:
+            try:
+                del self.parent.daemon_commands[k]
+            except KeyError:
```
Comment (cculianu) [inline lib/plugins.py]: Actually.. can happen if above raised `RuntimeError` on collision.. so good to catch this. 👍

File: lib/plugins.py [hunk 4]
```diff
@@ -545,6 +572,15 @@ def __init__(self, parent, config, name):
...
                 self._hooks_i_registered.append((aname,func))
 
+        # collect names of all class attributes with ._is_daemon_command
+        self._daemon_commands = tuple(attrname for attrname in dir(type(self))
+                                      if getattr(getattr(type(self),attrname), '_is_daemon_command',False))
+        # we don't allow conflicting definitions of daemon command (between different plugins)
+        collisions = set(self._daemon_commands).intersection(self.parent.daemon_commands)
+        if collisions:
+            raise RuntimeError('colliding daemon command names in plugin', collisions)
+        self.parent.daemon_commands.update({cmdname:getattr(self,cmdname) for cmdname in self._daemon_commands})
```
Comment (cculianu) [inline lib/plugins.py]: Hmm. I'd suggest writing to the `self._daemon_commands` dict once you have "committed" the commands to the parent dict and not before (before you can use a temp dict) --  This way if you raised `RuntimeError` above, your instance `dict` will be empty (which is really what you want). Thus, the invariant "if a command is in my local dict, it is also in the global one" is preserved.

File: lib/plugins.py (hunk 2 above)
Comment (markblundeberg) [reply lib/plugins.py]: Ah this is just a list of strings so it's no issue.

File: lib/plugins.py (hunk 4 above)
Comment (cculianu) [reply lib/plugins.py]: Additionally: If  `RuntimeError` was raised above and then this plugin is closed -- it will *clobber* the colliding plugin's registered commands as an unintended side-effect.

So I would suggest either clearing the dict before raising the exception, or writing to a temporary dict and then committing it to a final dict later.

Comment (markblundeberg) [reply lib/plugins.py]: If `RuntimeError` is raised, then there will be no instance so no problem. :-)

File: lib/plugins.py (hunk 2 above)
Comment (cculianu) [reply lib/plugins.py]: Oh, lol. ok :) Still tho.. what about defensive programming? Like kill the thing because who knows what evil lurks in other parts of the code where plugins can get "closed" twice, etc (that would be a bug in the plugin subsystem obviously but who knows..?)

File: lib/plugins.py (hunk 4 above)
Comment (cculianu) [reply lib/plugins.py]: Oh true, ha. This is a c'tor...

**Comment by cculianu** [comment]
I wonder if in a subsequent PR it would be worthwhile to expose the commands added by plugins to the argument parser somehow. Hmm.  Would require grabbing a reference to the global commands class and registering with the argparser when creating the plugin instance..  then cleaning up on close.. but may be a nice touch.

Can happen in a later PR and/or commit though....

**Comment by markblundeberg** [comment]
> I wonder if in a subsequent PR it would be worthwhile to expose the commands added by plugins to the argument parser somehow. Hmm. Would require grabbing a reference to the global commands class and registering with the argparser when creating the plugin instance.. then cleaning up on close.. but may be a nice touch.
> 
> Can happen in a later PR and/or commit though....

Yeah it's tricky since we need argparse to be initially active _without_ context: at very least it needs to be able to parse the commandline provided location of config file, and in the config file is where the to-be-loaded plugins are declared. Once config file is loaded then you can think about reparsing arguments and seeing if they belong to plugins. Once you start messing with argparse though, then you might lose the nice `--help` output.

**Comment by cculianu** [comment]
Yeah, I can see that. Catch-22.

File: lib/plugins.py (hunk 2 above)
Comment (cculianu) [reply lib/plugins.py]: Ok, let me try again since I fear this will get ignored: please please pretty please just empty that tuple when done with it?  A byte saved is a byte earned.  And also "mah invariants".  Etc.. Please?

Comment (markblundeberg) [reply lib/plugins.py]: Oh sorry, I missed this comment...
I still think it's fine... we don't delete most instance variables, just the ones that would cause a cycle.

Comment (cculianu) [reply lib/plugins.py]: true. I'm just a control freak.
